### PR TITLE
Initial work on formulating software metadata requirements and instructions

### PR DESCRIPTION
**Pull request by proycon**
This is an initial attempt to formulate software metadata requirements and instructions (#40). There's still a lot of TODOs pointing to ongoing discussions but I hope this clarifies the direction we're heading in and allows others to comment on it and provide the opportunity to steer in another direction where needed.

Direct link to the full text for ease of access: https://github.com/CLARIAH/clariah-plus/blob/software-metadata-requirements/requirements/software-metadata-requirements.md

### CODE REVIEW

File: requirements/software-metadata-requirements.md [hunk 1]
```diff
@@ -0,0 +1,187 @@
+# CLARIAH Software Metadata Requirements and Instructions
+
+## Introduction
+
+This document specifies requirements and instructions for software metadata for
+CLARIAH software and services. It is aimed at the developers of software and
+attempts to provide clear documentation.
+
+The main principle underlying the way we keep our software metadata is that
+**metadata is kept at the source and by the source of the software**. This
```
Comment (ddeboer) [inline requirements/software-metadata-requirements.md:10]: ‘by the source’ sounds a bit odd. Do you mean by the _maintainers_ of the software?

File: requirements/software-metadata-requirements.md [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+means that the tool developers themselves are responsible for providing
+accurate metadata and that the best place to keep this metadata is *with the
+source code* of the software, in a version-controlled repository.
+
+This is in line with current best-practices as most programming language
```
Comment (ddeboer) [inline requirements/software-metadata-requirements.md:15]: The approach aligns with current practices (using package definitions local to the programming language used) but which _best_ practices?

File: requirements/software-metadata-requirements.md [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+ecosystems already have their own ways of describing software metadata
+alongside the source code. Our second principle is to **avoid any duplication**
+so we make use of those existing metadata descriptions and convert them
+automatically to a unified [codemeta](https://codemeta.github.io/)
```
Comment (ddeboer) [inline requirements/software-metadata-requirements.md:19]:
```suggestion
automatically to a unified [CodeMeta](https://codemeta.github.io/)
```

File: requirements/software-metadata-requirements.md [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+representation, which is a linked open data vocabulary built on top of
+[schema.org](https://schema.org).
```
Comment (ddeboer) [inline requirements/software-metadata-requirements.md:21]:
```suggestion
[Schema.org](https://schema.org).
```

File: requirements/software-metadata-requirements.md [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+
+1. Adhere to the best practices in your programming language ecosystem and specify basic metadata in the way suited for your ecosystem.
+    * For Python this is a ``setup.py`` or ``pyproject.toml`` file
+    * For Java/Maven this is a ``pom.xml`` file
+    * For NodeJS/npm this is a ``package.json`` file
+    * Have a proper `README.md` file in all cases
+    * Have a proper `LICENSE.md` or `LICENSE` file in all cases
+    * All releases of your software should be accompanied with a git tag corresponding to the version number of the release. The harvester will always checkout the highest git tag (i.e. the latest version number) and harvest that instead of the git head of your master/main branch. The main/master branch will only be used as a fall-back in case you have no released versions at all yet.
+2. Run the `codemeta-harvester` on your project, it will attempts to automatically extract and convert metadata and produces a ``codemeta.json``
+    * **TODO: add instructions on how to run this**
```
Comment (ddeboer) [inline requirements/software-metadata-requirements.md:64]: Additionally, offer the validation as a web service.

**Review by ddeboer — COMMENTED** [review]
Good start! 

As a general observation, this still reads too much as a description of the status quo and its components (harvester etc.). Let’s make this document more to the point and focused on the question: how can developers publish their software metadata to the CLARIAH tool registry?

**Comment by proycon** [comment]
> As a general observation, this still reads too much as a description of the status quo and its components (harvester etc.). Let’s make this document more to the point and focused on the question: how can developers publish their software metadata to the CLARIAH tool registry?

That's a good point yeah, I'll rewrite some parts to have a clearer focus and document the pipeline itself elsewhere and link to it.

File: requirements/software-metadata-requirements.md (hunk 5 above)
Comment (proycon) [reply requirements/software-metadata-requirements.md:64]: We could do that yes. Currently the validation is just a component in the harvesting pipeline which runs in one of two ways:

* the harvester pipeline that runs periodically (https://tools.dev.clariah.nl), also provides feedback but at a delay (runs once a day).
* the harvest run locally by the user, there's a docker container image available so people can run it easily to generate their own `codemeta.json` (and inspect validation output)

I suppose an extra validation service might be something where you paste your `codemeta.json` and have it validated.

**Comment by proycon** [comment]
I reworked this document now to hopefully be more focused and to the point, it's closer to the other requirement documents now and uses the same RFC2119 language. I also filled more details as work is progressing.

**Comment by proycon** [comment]
If there are no objections, I'd like to merge what we have thus-far into the main branch, we can then continue from there in further incremental issues/PRs.

**Pull request merged**